Replace matrix-vector multiplication kernel with kernel generator

t4c1 opened the pull request:
## Summary
Replaces OpenCL kernel for matrix-vector multiplication with kernel generator. 

## Tests
No new tests. New code is tested as a part of OpenCL matrix multiply tests.

## Side Effects

Since `multiply.hpp` now includes `kernel_generator.hpp`, `binary_operation.hpp` can not include `multiply.hpp` anymore. This was solved by moving code for matrix-multiplying kernel generator expressions into `multiply.hpp`. Some headers that include `kernel_generator.hpp` must now also include `multiply.hpp`.

## Release notes
OpenCL kernel for matrix-vector multiply is replaced with kernel generator.

## Checklist

- [x] Math issue #1342 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

SteveBronder commented:
Would you mind throwing together a performance benchmark so we can see the diff here? Would be very cool if no change!

t4c1 commented:
Benchmark shows there form most shapes there is little to no difference, but for long vector and small matrix it goes up to 20%:
develop:
```
---------------------------------------------------------------------
Benchmark                              Time           CPU Iterations
---------------------------------------------------------------------
matvec_mul/1000/1000_mean         196773 ns     185904 ns       3446
matvec_mul/1000/1000_median       196227 ns     185904 ns       3446
matvec_mul/1000/1000_stddev         5788 ns       4780 ns       3446
matvec_mul/32/30000_mean         2395988 ns    2360491 ns        280
matvec_mul/32/30000_median       2392152 ns    2371652 ns        280
matvec_mul/32/30000_stddev          8483 ns      45942 ns        280
matvec_mul/30000/32_mean          143619 ns     138114 ns       4480
matvec_mul/30000/32_median        143802 ns     139509 ns       4480
matvec_mul/30000/32_stddev          2545 ns       4708 ns       4480
matvec_mul/20000/20000_mean     16124742 ns   16158537 ns         41
matvec_mul/20000/20000_median   16131880 ns   16006098 ns         41
matvec_mul/20000/20000_stddev      92101 ns     196798 ns         41
matvec_mul/32/600000_mean       48273999 ns   48325893 ns         14
matvec_mul/32/600000_median     48264200 ns   47991071 ns         14
matvec_mul/32/600000_stddev        30765 ns     753291 ns         14
matvec_mul/600000/32_mean        1547018 ns    1113281 ns        640
matvec_mul/600000/32_median      1543219 ns    1098633 ns        640
matvec_mul/600000/32_stddev        10755 ns     124594 ns        640
```
this PR:
```
---------------------------------------------------------------------
Benchmark                              Time           CPU Iterations
---------------------------------------------------------------------
matvec_mul/1000/1000_mean         209781 ns     201172 ns       3200
matvec_mul/1000/1000_median       209934 ns     202637 ns       3200
matvec_mul/1000/1000_stddev         3915 ns       8855 ns       3200
matvec_mul/32/30000_mean         2823423 ns    2830070 ns        249
matvec_mul/32/30000_median       2823331 ns    2823795 ns        249
matvec_mul/32/30000_stddev          4563 ns      19844 ns        249
matvec_mul/30000/32_mean          150718 ns     144071 ns       4978
matvec_mul/30000/32_median        151800 ns     144385 ns       4978
matvec_mul/30000/32_stddev          3483 ns       3121 ns       4978
matvec_mul/20000/20000_mean     15895895 ns   15891768 ns         41
matvec_mul/20000/20000_median   15899212 ns   16006098 ns         41
matvec_mul/20000/20000_stddev      10391 ns     184088 ns         41
matvec_mul/32/600000_mean       58439261 ns   58380682 ns         11
matvec_mul/32/600000_median     58436591 ns   58238636 ns         11
matvec_mul/32/600000_stddev        12216 ns     449187 ns         11
matvec_mul/600000/32_mean        1556244 ns    1159668 ns        640
matvec_mul/600000/32_median      1555521 ns    1135254 ns        640
matvec_mul/600000/32_stddev         3579 ns      86317 ns        640
```

I figured out what is likelz causing these. I will trz to optimise it.

SteveBronder commented:
icic, could you send me the benchmark code? I've found that doing `ClobberMemory` etc in specific spots changes the benchmark a lot

t4c1 commented:
Benchmark code is here: https://github.com/bstatcomp/perf-math/blob/kg_matvac_multiply/kg_matvec_multiply.cpp

I doubt there is any issue with benchmarking code, as I am reasonably sure which part of the kernel causes the slowdown. However, it turns out optimizing it would get quite complicated. I will think about it again, but right now I am leaning towards not doing the optimization.

t4c1 commented:
I failed to optimize the kernel for this case. I think it should not matter too much as this is quite inefficient use of GPU anyway (it runs only 32 threads). If we wanted to optimize this case we should add another kernel specifically for the case of multiplying short and wide matrix with a long vector.

SteveBronder commented:
I went through and ran these benchmarks for a larger span of numbers and it looks like our current mat mul is either faster or equivalent

![image](https://user-images.githubusercontent.com/5857231/79804407-f5e0ad80-8331-11ea-8d2d-e3acbd45fc0c.png)


cpp/R code and data for bench

https://gist.github.com/SteveBronder/77d3d798bab249ba978c4fb21e1bfba6

SteveBronder commented:
Though at lot of these are for p small matrices we could run this for more data sizes to see what's going on. We also don't even kick in the opencl stuff until at least 512 size matrices

t4c1 commented:
Those are square matrices? Seems more or less consistent with my benchmark. I think the difference is small enough that we could replace it. If you disagree I can leave the current kernel in. However I will need kernel generator implementation of matrix vector multiplication for GLMs.

t4c1 commented:
I took another shot at optimizing this. This time I eleiminated most of the slowdown. With this change I think we can really replace the old kernel. I also redid the benchmark:
![kg_matvec_mul](https://user-images.githubusercontent.com/7499452/80094825-37a66980-8567-11ea-8c61-6f5330f9c756.png)

Optimizations are mostly in `operation_cl`, `rowwise_reduction_` and `load_`. All other files were changed as I refactored functionality of the `view()`, `top_diagonal()` and `bottom_diagonal` into a single function `extreme_diagonals()`. By doing this I fixed a bug, that could in some combinations of operations cause result view to be wrong. The test for that is in `block_test.cpp`. Due to large amount of changes in `rowwise_reduction_` I also added an additional test for it.

syclik commented:
@SteveBronder, mind reviewing this one?

SteveBronder commented on stan/math/opencl/kernel_generator/rowwise_reduction.hpp:
```diff
@@ -16,6 +18,60 @@
+        broadcast.var_name = name_gen.generate();
+        generated.insert(&broadcast);
+
+        const auto& vec_t = broadcast.template get_arg<0>();
```
We reserve `_t` stuff for type traits. Maybe `vec_expr`

t4c1 replied:
fixed

SteveBronder commented on stan/math/opencl/kernel_generator/rowwise_reduction.hpp:
```diff
@@ -16,6 +18,60 @@
+
+template <typename Arg>
+struct matvec_mul_opt {
+  enum { is_possible = 0 };
```
Though it's internal I think it would be good to doc these so it's easy to see what `is_possible` etc does from reading it here

t4c1 replied:
done

SteveBronder commented on stan/math/opencl/kernel_generator/scalar.hpp:
```diff
@@ -88,22 +91,12 @@ class scalar_ : public operation_cl<scalar_<T>, T> {
-   */
-  inline int top_diagonal() const { return std::numeric_limits<int>::max(); }
+  inline std::pair<int, int> extreme_diagonals() const {
+    return {std::numeric_limits<int>::min(), std::numeric_limits<int>::max()};
```
I'm not really following the code here. If this is supposed to grab the extreme diagonal indices why is it returning the min and max of int types?

t4c1 replied:
The scalar expression does not have size. In a binary expression it is always broadcast in both dimensions to match the other argument. So we don't know what exactely its extreme diagonals are. By returning min and max  int we are saying that regardless of size it is broadcast to it covers whole matrix - there are no zero parts. Broadcasting also returns min or max to similar effect.

SteveBronder commented on stan/math/opencl/kernel_generator/select.hpp:
```diff
@@ -103,14 +106,22 @@ class select_ : public operation_cl<select_<T_condition, T_then, T_else>,
-    matrix_cl_view then_view = this->template get_arg<1>().view();
-    matrix_cl_view else_view = this->template get_arg<2>().view();
-    return both(either(then_view, else_view), both(condition_view, then_view));
+  inline std::pair<int, int> extreme_diagonals() const {
+    using std::max;
+    using std::min;
+    std::pair<int, int> condition_diags
+        = this->template get_arg<0>().extreme_diagonals();
+    std::pair<int, int> then_diags
+        = this->template get_arg<1>().extreme_diagonals();
+    std::pair<int, int> else_diags
+        = this->template get_arg<2>().extreme_diagonals();
+    return {max(min(then_diags.first, else_diags.first),
+                min(condition_diags.first, else_diags.first)),
+            min(max(then_diags.second, else_diags.second),
+                max(condition_diags.second, else_diags.second))};
```
This `extreme_diagonals` seems not trivial in it's logic. Am I reading this right that you want the lowest to be the max of the minimum values of the diagonal elements of the first 3 args and the highest to be the min of the maximum values of the diagonal elements of the first 3 args. It could be useful to add docs here since the logic is very particular

t4c1 replied:
Yeah it is definitely not trivial. Added a comment

t4c1 commented:
@SteveBronder This is ready for next review.
